app_dart/lib/src/request_handlers/check_for_waiting_pull_requests.dart

```diff
@@ -0,0 +1,280 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert' show jsonEncode;
+
+import 'package:appengine/appengine.dart';
+import 'package:cocoon_service/src/request_handling/exceptions.dart';
+import 'package:graphql/client.dart';
+import 'package:meta/meta.dart';
+
+import '../datastore/cocoon_config.dart';
+import '../foundation/providers.dart';
+import '../foundation/typedefs.dart';
+import '../request_handling/api_request_handler.dart';
+import '../request_handling/authentication.dart';
+import '../request_handling/body.dart';
+
+import 'check_for_waiting_pull_requests_queries.dart';
+
+@immutable
+class CheckForWaitingPullRequests extends ApiRequestHandler<Body> {
+  const CheckForWaitingPullRequests(
+    Config config,
+    AuthenticationProvider authenticationProvider, {
+    @visibleForTesting LoggingProvider loggingProvider,
+  })  : loggingProvider = loggingProvider ?? Providers.serviceScopeLogger,
+        super(config: config, authenticationProvider: authenticationProvider);
+
+  final LoggingProvider loggingProvider;
+
+  @override
+  Future<Body> get() async {
+    final Logging log = loggingProvider();
+    final GraphQLClient client = await config.createGitHubGraphQLClient();
+
+    await _checkPRs('flutter', 'flutter', log, client);
+    await _checkPRs('flutter', 'engine', log, client);
+
+    return Body.empty;
+  }
+
+  Future<void> _checkPRs(
+    String owner,
+    String name,
+    Logging log,
+    GraphQLClient client,
+  ) async {
+    bool hasMerged = false;
+    final Map<String, dynamic> data = await _queryGraphQL(
+      owner,
+      name,
+      log,
+      client,
+    );
+    for (_AutoMergeQueryResult queryResult in _parseQueryData(data)) {
+      if (!hasMerged && queryResult.shouldMerge) {
+        hasMerged = await _mergePullRequest(
+          queryResult.graphQLId,
+          queryResult.sha,
+          log,
+          client,
+        );
+      } else if (queryResult.shouldRemoveLabel) {
+        await _removeLabel(
+          queryResult.graphQLId,
+          queryResult.removalMessage,
+          queryResult.labelId,
+          client,
+        );
+      }
+    }
+  }
+
+  Future<Map<String, dynamic>> _queryGraphQL(
+    String owner,
+    String name,
+    Logging log,
+    GraphQLClient client,
+  ) async {
+    final String labelName = await config.waitingForTreeToGoGreenLabelName;
```

Comment:
Move this to parent function would save it from waiting every time.

Reply:
This should be fast after the first time - datastore does some caching I believe, right?

I could move it up the call stack but then we're passing the label around and these parameter lists are already fairly long.  That could probably be improved though by making this class not be immutable..  I'll do that instead.

Reply:
Actually I can't do that because the parent class is immutable.